[MRG] SSS: Add Fine Calibration and Cross-talk Correction

wronk opened the pull request:
Add improved calibration to the SSS functionality. This should increase the SSS shielding factor by ~10x making it useful for actual data processing.

* [x] Add option to use sensor geometry from `sss_cal.dat` files instead of `raw`
* [x] Quantify difference in geometry and send to `logger` along with fine calibration file used
* [x] Add point-like magnetometers (X, Y, and Z) to gradiometer calcs
* [x] Add sensor imbalance weights to gradiometer calcs and calibration coeffs to magnetometer calcs
* [x] Add cross-talk correction
* [x] Add tests for most/all of the above
* [x] Test shielding factor using empty room
* [x] Head position translation
* [x] Verify against Elekta's code
* [x] Add warning about T1/T2 magnetometers if present and fine cal not used (#2502)
* [x] Ensure processing history documents cross-talk and fine cal

Closes #2502.

This will require an update of the mne-testing-dataset

agramfort commented:
I don't think we want in the repo the file from a specific site. To me it
should be passed as parameter and maybe remembered in the MNE config file.

wronk commented:
@agramfort, I was going to change it later for and get everything running first, but might as well do it right the first time. We were discussing what the best way to handle the `.dat` files in the issue thread. Here's the summary as I interpreted it:

@Eric89GXL's suggestion:
Automatically download the `.dat` files separately from the MNE package. For now, user might need to specify the correct one, but we can try to auto detect the correct `.dat` file from `.raw` files eventually.  Where exactly would we store it if we go this route? Would it be similar to the sample and testing datasets where you auto-download and configure a path the first time the data is needed?

@dgwakeman suggestion:
Construct a symlink to the `.dat` file the first time fine-calibration is used and optionally construct your own fine-calibration files. (This second point is possible from ERM data, but I don't see it as a high priority atm.) Alternatively, read the fine-cal file from the MEG system's files by SSH'ing in. There is a standard location on those computers, but there will be issues if someone doesn't have access or they don't work at an MEG site.

dgwakeman commented:
I think the most accurate/safest thing we could do would be to on first run ask the user to provide the path to the `.dat` file and code this in the MNE config etc., and generate a dictionary that we use that pairs with the `machid` from the system (we shouldn't deal with any ssh'ing all sites that I am aware of have maxfilter available on linux machines besides the acquisition computer). We can then check that the `machid` matches the `.dat` file dictionary. This is not super safe, but is better than the alternative. It also provides an option for having multiple entries in the dictionary, when someone is analyzing data from multiple sites.

This option has the following issues:
False Positives: a site changes their acquisition computer and then has a new `machid`, but the same calibration file.

    To deal with this we would just need to add the new `machid` and point it to the same `.dat`
False Negatives: a site replaces several MEG sensors, and doesn't update the computer. 

    If we wanted, we could try to add some date information to check the validity, but that sounds complicated to me. and not worth the extra cost (at least for now).

wronk commented:
That sounds like a better alternative to SSH'ing, but how should we make the `.dat` file available to the user?

(For the moment, I'll simply ask for the filepath to get going on the actual implementation, but let's continue this convo.)

Eric89GXL commented:
@wronk I have come around to @dgwakeman's plan actually, and I really like the `machid` plan. But I think that instead of making a symlink, we might want to store the file in `.mne/maxfilter` or so. Seems nicer for persistence to me. @dgwakeman sound okay to you?

dgwakeman commented:
@Eric89GXL, Not too concerned about the details. Personally, I don't think it belongs in `~/.mne/`, but it should be where MNE-python is installed or some related site local config dir. In this way, experts on site could configure it for all users (this would also reduce the chance of user error), but I'm not sure how much extra work that is. I am happy with whatever you guys thinks makes the most sense.

Eric89GXL commented:
.mne is actually where we already store bits of specific data, so it mostly
fits with what you are thinking

wronk commented:
Sounds like a plan for the moment, and I'll ask again once I start implementing that part.

Different question: One part of fine-calibration involves setting slightly more accurate positions and orientations of the MEG sensors, which will change the `'coil_trans'` and `'loc'` keys within each coil dict in `raw.info['chs']`. Should I supplement this channel information by making keys like `'coil_trans_fine'` and `'loc_fine'` and pass around a `fine_cal` flag to know when to use fine calibration or overwrite that information in `raw.info`? I'm leaning towards the former but wanted to check.

Eric89GXL commented:
Does MaxFilter overwrite it in its output files?

wronk commented:
Yes, just checked it out and the returned file has different coil location information. Should I follow suit?

Eric89GXL commented:
Yep, please do

dgwakeman commented:
@Eric89GXL, so does mne-python generate its own `.mne` where it is installed, or are you talking about `~`?

agramfort commented:
as a sidenote we should prevent people to use a certain fine calibration file when the file was acquired on a certain machine. I remember asking people at MGH to run maxfilter on files acquired in Germany... I had no idea it was a bad idea...

agramfort commented:
+1 to put the files in .mne in home dir but be careful not to write anything in home dir during the tests as it will break debian packaging

Eric89GXL commented:
Yeah it's in the home directory @dgwakeman

Yes be careful about writing, follow what the forward code does and you
should be safe

dgwakeman commented:
-1 on home dir. IMO, this should only be for user specific files: color preferences, user specific layouts etc. Plus now you have tons of copies of the same thing in every home users directory, and more experienced users can't help other users by setting up the local sites information in the initial dictionary set.

Eric89GXL commented:
Oh, you're thinking from the perspective of a bunch of users using the same machine, in which case it would be less convenient. I'm thinking from the perspective of many users, each on their own machine. I think this is probably the majority of use cases ideally -- a major restriction on the current MaxFilter is that you have to run it on specific machines to have the proper setup, and we're going to alleviate that limitation hopefully...

I think I don't fully understand your proposal, then. You would have a dict that would get saved to each user's machine, instead? Then there are still copies everywhere that need to be maintained, but the dictionary pointers are redundant across machines instead of the files being redundant. In my mind those aren't sufficiently different to warrant making the individual-user case (where we store the files for them) less convenient.

dgwakeman commented:
Let's first attack a few key misconceptions:
1. the sss_cal.dat file is virtually never installed locally on a single computer (I have never heard of this happening at any site). It is always installed on a network file system, in some shared accessible directory. 
2. All licensing is handled by the maxfilter binary, which doesn't concern us at all (that is why some machines can't run the binary, but the cal file is not local!)

Most sites have a central mne-python installation, which most users use for analysis (with the exception of a handful of advanced users/developers, who roll their own). Let us call the location of that installation $MNE_PYTHON_ROOT. Somewhere around there is where I propose saving the dictionary and possibly hard linking/copying the sss_cal.dat files. In this way, each user won't have to add the sss_cal.dat: a potentially hazardous step.

make sense?

Eric89GXL commented:
> Most sites have a central mne-python installation, which most users use for analysis

Is that true? I assumed that was not the case, actually.

> Let us call the location of that installation 

One problem is that this location is often not writable by users, and this is true even more so for shared machines. This means if a user wants to have their MF cal file stored -- e.g. if they start analyzing data from another site -- they now have to go through someone with `root` permissions to get the desired behavior.

Eric89GXL commented:
What could work is having users set a `MNE_CAL_PATH` variable that defaults to `.mne/`. You just have those users do `mne.set_config('MNE_CAL_PATH', '/shared/wherever')` in your use case, giving you all the control you need. Make sense?

mainakjas commented:
On Tue, Aug 4, 2015 at 8:32 PM, Eric Larson <notifications@github.com>
wrote:

> Most sites have a central mne-python installation, which most users use
> for analysis
>
> Is that true? I assumed that was not the case, actually.
>
I know that in Aalto that is in fact true. Don't know about other locations.

> Let us call the location of that installation
>
> One problem is that this location is often not writable by users, and this
> is true even more so for shared machines. This means if a user wants to
> have their MF cal file stored -- e.g. if they start analyzing data from
> another site -- they now have to go through someone with root permissions
> to get the desired behavior.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2374#issuecomment-127642707>
> .
>

dgwakeman commented:
It is true at Martinos, and at Cambridge, and I also believe at Oxford.

Why have it default to `~/.mne`? Maybe you could do a secondary search there and issue a warning, but having a central location for these things is useful, and reduces potential problems. In these instances there is usually someone managing this installation, and that person can update these files. (but I can see why we may want them to be outside the `$MNE_PYTHON_ROOT`, something like: `$MNE_PYTHON_ROOT/../.mne`

Eric89GXL commented:
Interesting, I didn't realize there were so many site-wide installs going on!

I think `~/.mne` is still the most sensible default, though. Maybe we can make the user enter a location the first time we search the path, too, to ensure they have an opportunity to enter the site-specific path. 

I don't know what a better default would be. Something like `$/MNE_PYTHON_ROOT/../.mne` is not a good choice, because it will fail on many installs (site-wide and user-specific, actually), won't work for all sites anyway, etc. It seems cleanest to have users of such distributions do `mne.set_config(...)` to get the behavior they need for their specific site, or have another first-use prompt that defaults to `~/.mne`. This would also be consistent with how we have stored and accessed files up to this point, and general mne-python schemes.

dgwakeman commented:
The constant filling of ~/.mne is incredibly frustrating to me and I have to repeatedly delete that entire directory: and backup the stuff I actually want there. Most sites have small strict limits (1-2GB) to home directories, and mne-python seems to put a ton of junk there that seems to have no purpose:
ls -lh ~.mne/tables/
total 46M
-rw-rw-r-- 1 dwakeman dwakeman  23M Jun 10  2015 legder_100_20000.bin
-rw-rw-r-- 1 dwakeman dwakeman  12M Jun 10  2015 legder_50_20000.bin
-rw-rw-r-- 1 dwakeman dwakeman 7.7M Jun 10  2015 legval_100_20000.bin
-rw-rw-r-- 1 dwakeman dwakeman 3.9M Jun 10  2015 legval_50_20000.bin

Here's 50megs of crap, that I don't want in my home directory, and it just did it without asking or anything.

Eric89GXL commented:
FYI those files are LUTs used to substantially increase forward solution
calculation speed.

I think you can tell mne to use a different directory, and there is also a
key MNE_DONTUSE_HOME or so, that you can set to prevent it from writing
there at all. Feel free to open a PR to improve or  document this in case
anyone you suspect others are similarly frustrated.

However, I think for the majority of users this behavior is reasonable, and
the tradeoff for storage is worth the benefits, so I am -1 on changing this
default behavior.

wronk commented:
After talking more with @staulu, overwriting the channel positions is more complicated than I thought. If fine calibration is on in Elekta's Maxfilter, it will compute multipolar moment bases using both the positions and orientations supplied in the `sss_cal.dat` file (as expected). However, the raw file you get back is a hybrid containing coil positions as in the standard calibration and coil orientations as in the fine calibration.

@staulu said this was a matter of opinion, so we should discuss what we do. I'm most in favor of either overwriting position and orientation info or including fine calibration versions alongside the original position/orientation. The keys in question are `raw.info['loc']` and `raw.info['coil_trans']`.

Eric89GXL commented:
I don't think we want new entries, so we should either overwrite, or not. My +1 is for overwriting both, assuming that both values are more accurate. Any idea whether they are both more accurate? Were positions not overwritten e.g. so that sensor layouts could be properly detected or something?

wronk commented:
Yes, they should both be more accurate. I believe positions were left as is in an attempt to keep the user less confused.

I think we can keep this pretty clear though in the documentation. I'll move ahead with overwriting (since that's what it does now anyways :p)

Eric89GXL commented:
Can you at some point roughly quantify the differences we are talking? 100 microns? 1 mm? median and max differences would be interesting. In fact, you might as well `logger.info` what the mean and std dev adjustment amount resulting from the fine cal is.

wronk commented:
Ya, that would be useful -- added to the checklist

wronk commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +536,97 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+                str(np.std(shift_arr, 0)))
+    logger.info('Max change in sensor location and orientation calculated as\n'
+                ' (np.max(np.abs(shift))): ' +
+                str(np.max(np.abs(shift_arr), 0)))
```
@Eric89GXL, see if this is what you had in mind for logging. The output isn't very pretty (as just a 1x14 array of values), but it does match the format for the `sss_cal.dat` files. Is that okay, or should I format this logging output differently?

dgwakeman replied:
Are people concerned that this will make people worry about what this is doing? and we will get a lot of emails saying, "wait why is it changing the positions of the sensors?"

Eric89GXL replied:
I'd call it an "adjustment" instead of a change, it will make people more comfortable. I'm fine with getting those emails if they come, it's better that people know what's happening to their data in my mind.

@wronk how about this:
```
"Fine calibration adjusted coil positions by (mm): <mean> ± <std> SD (max: <max>)"
```
Why are there 14 values? I'd just operate on the flattened array.

wronk replied:
There are positions and orientations with different scales/meanings, so I kept everything in the same format.

Eric89GXL replied:
For the first pass I wouldn't worry about the orientations, just give stats on the positions. Those should all be in `mm` and thus can be collapsed, right?

Eric89GXL replied:
We aren't trying to give users a full picture, just some sense of how much things are likely changing, and the position changes should convey that clearly enough hopefully.

wronk replied:
Sure, I'll update to give a stats on x/y/z position changes.

Eric89GXL replied:
I wouldn't separate the coordinate directions, it would still be more information than necessary. You should describe the total distance moved

staulu commented:
Just a quick clarification: The positions stay intact in the fine calibration procedure. The only geometrical factor that changes are the channel orientations. This is because SSS is much more sensitive to inaccuracies in the channel orientations.

Eric89GXL commented:
Ahh, I see. @wronk I guess you could figure out the angular change that occurred (e.g. via `arccos` of the dot product of the original orientation vector and the new one), and `logger.info` that instead.

wronk commented:
Sure, I'll do that. I didn't realize that positions weren't modified...

wronk commented:
@staulu, what units do the orientations have in the `sss_cal.dat` file? The look to be in the range (-1, 1).

@Eric89GXL, do we want to report this in a different unit or keep it the same as defined in the `.dat` file? It seems like not converting will make it easier for someone to compare this difference with the `.dat` file itself and avoid someone having to do their own back conversion.

Eric89GXL commented:
Those might be quaternions...?

It's better to put it in human readable values, i.e. degrees. The idea is
to give people some simple, easy to understand / ballpark measure, not a
compete report. I highly doubt anyone will check, and anyone who does will
be motivated enough to do the required math.

staulu commented:
The orientations in the sss_cal.dat file are just unit vectors in the Cartesian coordinate system. Definitely the clearest way to illustrate their change from the default is to calculate the angle in degrees between, e.g., the normal vector before and after fine calibration.

wronk commented:
Ah, of course. Logging for this is now updated.

Eric89GXL commented:
So where are we on this now? Are you checking results against the last version you benchmarked, but with the `-cal <file>` option turned on?

wronk commented:
Ya, currently checking against Elekta's maxfilter and going to review the code with Samu today.

Are we generally settled on writing the `.dat` file to ~/.mne/maxwell_filter and checking `machid`s?

dgwakeman commented:
@agramfort said:

    Regarding cal file I would use a path in the mne config file that would point on sites to the system
    location. Let's avoid multiple copies of this file and let's add to the logger the file used when
    running maxwell_filter in python.

And I would say we should add the dictionary to the config file as well.

agramfort commented:
yeah don't copy the file but use just a path and raise an error if file is
not there.

wronk commented:
If it's alright, I think I'm going to shelve the storing of the `sss_cal.dat` file for now and force the user to pass a file name. I'm running low on time and feel like this will take a while given my inexperience with saving/storing things correctly in a package. I'd rather spend the time trying to finish this and then get tSSS up and running.

Eric89GXL commented:
+1 from me, we can deal with it later

agramfort commented:
fair enough

Eric89GXL commented on mne/data/coil_def_Elekta.dat:
```diff
@@ -39,6 +39,12 @@
 #			2	accurate
 #
 #
+1   2000    2   1  0.000e+00  0.000e+00	"Point magnetometer, x-normal"
```
is `2000` actually a named `FIFF` constant? Can you check? If so, we should add it to `mne.io.constants`

wronk replied:
Currently, 
`FIFF.FIFFV_COIL_POINT_MAGNETOMETER = 2000  # Simple point magnetometer`, which refers to a point magnetometer with a normal vector in the z-direction. I also want to have point magnetometers in the x and y directions. Should I shuffle everything so 
2000 refers to x
2001 refers to y
2002 refers to z?

Or would it be better to append y and z? Leaning towards former option if it won't cause errors.

wronk replied:
Although currently `FIFF.FIFFV_COIL_AXIAL_GRAD_5CM     = 2001  # Generic axial gradiometer`...

Eric89GXL replied:
Argh. Where does it say that the 2000 has a normal vector in the Z direction? That is contrary to what MaxFilter / `coil_def` is saying, right?

Eric89GXL replied:
Numbers can be used multiple times, unfortunately. I wonder if this is a case where they did accidentally.

wronk replied:
The z direction is fine, that's how I would've expected it to be defined. The final three numbers of the current MaxFilter point magnetometer definition (corresponding to the three normals) are 0, 0, 1.

Eric89GXL replied:
> which refers to a point magnetometer with a normal vector in the z-direction

That's what I'm asking about -- where is it defined in the `z` direction? In this file it looks like it's in the `x` direction, and the comment suggests that, too.

wronk replied:
Maybe we're talking about different files... In the original `coil_def.dat` file:
>1   2000    0   1  0.000e+00  0.000e+00	"Point magnetometer"
> 1.0000  0.000e+00  0.000e+00  0.000e+00  0.000  0.000  1.000

`coil_def_elekta.dat` has them defined differently (2000 is x, 2001, is y, 2003 is z), but it should probably be consistent with `coil_def.dat`, right?

Eric89GXL replied:
@mshamalainen and @staulu any idea why these point magneteter definitions
would differ between MNE and Elekta? It would be nice if we could reconcile
them.

wronk replied:
Sorry, I defined them that way on my own. These aren't actually ported from Elekta's coil definitions or anything. I just added them to the `coil_def_elekta.dat` for simplicity.

Eric89GXL replied:
Oh, then yes make them match :)

wronk replied:
Should I do that by also modifying `coil_def.dat`? I can make 2000, 2001, 2002 match x, y, z that way

Eric89GXL replied:
No. `coil_def.dat` is pulled from MNE-C, since you're making the new values here, we should change them here.

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +537,107 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+    # Initialize internal multipolar moment space for point-like magnetometers
+    S_in_all = np.zeros((len(info['chs']), (int_order + 1) ** 2 - 1))
+    # Coil_type values for x, y, z point magnetometers
+    pt_types = [2000, 2001, 2002]
```
...and then change these to be `FIFF.WHATEVER`

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +537,107 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+    with open(fine_cal_fname, 'r') as fid:
+        cal_chans = []
+        lines = fid.readlines()
+        lines = lines[::-1]
```
comment why you omit, probably `  # blank line`

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +537,107 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+        cal_chans = []
+        lines = fid.readlines()
+        lines = lines[::-1]
+        while len(lines) > 0:
```
boo unnecessary `while` loop

`for line in lines` is clearer when you know the number of lines you're going to do. Save a couple lines by skipping the above lines and just doing:
```
for line in fid.readlines()[:-1]:
```

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +537,107 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+        lines = lines[::-1]
+        while len(lines) > 0:
+            line = lines.pop()
+            if line[0] != '#':
```
can save indentation of the following by doing `if line[0] == '#': continue` if you want

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +537,107 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+                ch = int(vals[0])
+                ch_name = 'MEG' + '%04d' % ch  # Zero-pad names to 4 characters
+                loc = vals[1:13]
+                imbalance = vals[13:]
```
FYI this will keep all of the original `vals` arrays in memory, because slices are always copies, and thus `vals` cannot be cleared from memory. It's fine here because there are so few so we don't need to worry about it, but something to keep in mind for future work. If you really want to be clean, just add `.copy()` to the end of these, and the garbage collector will do its work.

wronk commented:
Data filtered with fine calibration on here is not as close to Elekta's version as I'd expect. I'll keep looking for a bug and work probably with @staulu when he's back in the country next week. In the meantime, I'll get started on tSSS.

Eric89GXL commented:
Sounds good to me

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -90,25 +103,55 @@ def maxwell_filter(raw, origin=(0, 0, 40), int_order=8, ext_order=3,
-    coil_scale[np.array([coil['coil_class'] == 1.0 for coil in coils])] = 100.
+    magscale = 100.
+    coil_scale = np.ones((len(picks), 1))
+    coil_scale[np.array([coil['coil_class'] == 1.0
```
you shouln't need to do the `np.array(...)` here, a list should be fine (`numpy` should automatically cast to array, I think)

wronk replied:
Doesn't seem to be working for me :/

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -232,7 +281,6 @@ def _sss_basis(origin, coils, int_order, ext_order):
 
         # Scale magnetometers and normalize basis vectors to unity magnitude
```
remove the comment about normalizing here, since you've moved it

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -183,19 +227,24 @@ def _sss_basis(origin, coils, int_order, ext_order):
+
+    Notes
+    -----
+    Incorporates magnetometer scaling factor. Does not normalize spaces though
```
delete `though`, add period

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -90,25 +103,55 @@ def maxwell_filter(raw, origin=(0, 0, 40), int_order=8, ext_order=3,
     raw_sss = _update_sss_info(raw.copy(), origin, int_order, ext_order,
                                data.shape[0])
-    raw_sss._data[picks, :] = recon / coil_scale[:, np.newaxis]
+    raw_sss._data[picks, :] = recon / coil_scale
```
to save memory:
```
raw_sss._data[picks] = recon
raw_sss._data[picks] /= coil_scale
```
otherwise a temporary array the same size as `recon` gets created.

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +553,113 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+        err_msg = '''Number of channels in fine calibration file (%i) does not
+            equal number of channels in info (%i)''' % (len(cal_chans),
+                                                        info['nchan'])
+        raise RuntimeError(err_msg)
```
as before, combine into one line please

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +553,113 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+    """Read sensor locations and calib. coeffs from fine calibration file."""
+
+    logger.info('Fine calibration file used in Maxwell filter: ' +
+                fine_cal_fname)
```
This `logger.iinfo` should probably actually be in `maxwell_filter`, not in the fine cal reading. That way this function could be used by something (anything) else that needs to read that type of file. Think orthogonality and hopefully it makes sense.

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +553,113 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+    return cal_chans
+
+
+def _update_sens_geometry(info, cal_chans):
```
might as well spell out `sensor` instead of `sens` here

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +553,113 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+
+    # Log quantification of sensor changes
+    logger.info('Fine calibration adjusted coil positions by (mean ± std in '
+                'degrees): %0.5f ± %0.5f (max: %0.5f)' %
```
`0.1f` is probably good enough

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +553,113 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+        info['chs'][ch_ind]['loc'] = cal_chan['loc']
+
+    ang_shift[ang_shift > 1.] = 1.  # Deal with unreasonable values
+    ang_shift = np.arccos(ang_shift) * (180. / np.pi)  # Convert to degrees
```
might as well use `np.rad2deg(...)` instead of manually doing it yourself

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +553,113 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+        # Adjust channel orientation with those from fine calibration
+        info['chs'][ch_ind]['loc'] = cal_chan['loc']
+
+    ang_shift[ang_shift > 1.] = 1.  # Deal with unreasonable values
```
Do you know why this can occur? Is it like 1.00000001324 or something where it's numerical precision? If so, say it's due to numerical precision

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +553,113 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+                FIFF.FIFFV_COIL_POINT_MAGNETOMETER]
+
+    # Loop over all coordinate directions desired and create point mags
+    for dir_ind in np.arange(imbalance.shape[1]):
```
this can just be `range` instead of `np.arange`, on Py3k it's slightly more efficient (esp. for large numbers of points) since it's a generator instead of a list. Not a big difference here since it's 1 or 3, but worth changing to develop good habits.

wronk replied:
Did not know that, thanks for the tip :)

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +553,113 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+        S_in_add, S_out_add = _sss_basis(origin, coils_add, int_order,
+                                         ext_order)
+        # Scale spaces by gradiometer imbalance
+        S_in_all += S_in_add * imbalance[:, dir_ind].reshape(-1, 1)
```
the reshape is more vague / less common than doing `..., dir_ind][:, np.newaxis]`

Eric89GXL commented on mne/preprocessing/tests/test_maxwell.py:
```diff
@@ -25,11 +25,24 @@
+fine_cal_fname = op.join(data_path, 'SSS', 'sss_cal_3053.dat')
+fine_cal_fname_3d = op.join(data_path, 'SSS', 'sss_cal_3053_3d.dat')
+
+int_order, ext_order = 8, 3
```
since these are the defaults anyway, how about we remove them?

wronk replied:
I use these vars to test some of the more fundamental functions that don't use these values as defaults. (See lines 71-76)

Eric89GXL commented on mne/preprocessing/tests/test_maxwell.py:
```diff
@@ -25,11 +25,24 @@
+    Notes
+    -----
+    Elekta MaxFilter uses single precision, so expect filtered results to
+    differ slightly.
```
By convention we usually don't add notes to these functions, so just make this a `#` comment instead

Eric89GXL commented on mne/preprocessing/tests/test_maxwell.py:
```diff
@@ -157,4 +163,66 @@ def test_maxwell_filter_additional():
+                                                       for coil in all_coils]]
+
+    # Test 1D SSS fine calibration
+    raw_sss = maxwell.maxwell_filter(raw, origin=[0., 0., 40.],
```
to save some lines, do `from mne.preprocessing import maxwell_filter` at the top. Same with the private functions, you can do `from mne.preprocessing.maxwell import ...`. It's what we usually do anyway.

Eric89GXL commented on mne/preprocessing/tests/test_maxwell.py:
```diff
@@ -157,4 +163,66 @@ def test_maxwell_filter_additional():
+    assert_true(np.mean(bench_rms / error_rms) >= 100, 'SNR < 100')
+
+
+# TODO: Eventually add simulation tests mirroring Taulu's original papers
```
can you add a NIH Pubmed link here, or some other relevant link to the paper/papers?

Eric89GXL commented:
Code in general looks very good, I just had some minor tweak suggestions. Hopefully you can figure out the difference without too much trouble.

Travis also has some style complaints:

https://travis-ci.org/mne-tools/mne-python/jobs/74651744#L2471

This test differs by an order of magnitude, are you sure the bug is subtle?

https://travis-ci.org/mne-tools/mne-python/jobs/74651743#L2354

wronk commented:
There's a good chance the bug is something small leading to a big error. When I fixed that 2-3 line pseudoinverse problem we were having with the original SSS work, the error jumped down by 2 orders of magnitude.

`calib_coeff` is only stored in `cal_chans`, which is a list of channels I pass around independently of `info`.

Eric89GXL commented:
@wronk are you going to try rebasing this one today? If not, I might be able to add a cross-talk cancellation PR since it should be very simple. Then after that's in you could rebase this one. Or I'm happy to rebase this one for you :)

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +557,112 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+        info['chs'][ch_ind]['loc'] = cal_chan['loc']
+
+    # Deal with numerical precision giving values slightly more than 1.
+    ang_shift[ang_shift > 1.] = 1.
```
I think you also need `ang_shift[ang_shift < 1.] = 1.`. But really you can just do this:
```
np.clip(ang_shift, -1., 1., ang_shift)
```

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -505,3 +557,112 @@ def _update_sss_info(raw, origin, int_order, ext_order, nsens):
+
+    # Initialize internal multipolar moment space for point-like magnetometers
+    S_in_all = np.zeros((len(info['chs']), (int_order + 1) ** 2 - 1))
+    S_out_all = np.zeros((len(info['chs']), (ext_order + 1) ** 2 - 1))
```
more DRY to use the moment-number functions to get the `n_in` and `n_out`

wronk commented:
@Eric89GXL, ya, I'll get going on that rebase now

wronk commented:
@Eric89GXL, rebased

Eric89GXL commented:
Can I push some commits to fix a couple of bugs? I'm also thinking of adding the cross-talk check here since it should be so simple.

Eric89GXL commented:
or cross-talk calculation rather

wronk commented:
Yep, go for it

wronk commented:
I found one bug where not all the channel information was getting overwritten. SNR is ~10 at this point, so it's on the right track.

Eric89GXL commented:
Great. It should still be higher though right?

wronk commented:
Yep, I'm hoping get it above 250 like the all the other functionality PRs.

Eric89GXL commented:
Do you have more ideas for how to potentially fix it, or are you out of ideas?

I'll probably try to do the cross-talk correction in the next day or two and push to this branch.

wronk commented:
I do have a few more ideas and will probably try meet with Samu this week
if none of those pan out.

Just give me a heads up on the ctc and I'll be sure to commit what I have.

On Mon, Oct 5, 2015 at 4:26 PM, Eric Larson <notifications@github.com>
wrote:

> Do you have more ideas for how to potentially fix it, or are you out of
> ideas?
>
> I'll probably try to do the cross-talk correction in the next day or two
> and push to this branch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2374#issuecomment-145697255>
> .
>

Eric89GXL commented:
Will do. I'll probably add the warning about T1/T2 while I'm at it.

Eric89GXL commented:
I added to the top comment that we need to ensure that we respect the processing history. I think that if SSS has already been performed, we should (for now) throw an error. We should also not redo fine cal or cross-talk if it has been done already. From the MF manual:

> Fine-calibration adjustment is not performed if the fine-calibration file is
> not found, or the processing history of input_file.fif already includes the
> fine-calibration. Likewise the cross-talk correction is not done if the crosstalk
> matrix file is not found, or the processing history of input_file.fif
> already includes the correction

If we never allow redoing `maxwell_filter` if it has already been done this is less of an issue for us, but we should at least make sure that files we create are read properly by `maxfilter` such that cross-talk and fine cal (or even standard SSS unless `-force` is on) are not redone if they have been processed with our utility.

Eric89GXL commented on mne/data/coil_def_Elekta.dat:
```diff
@@ -12,8 +12,7 @@
 #	1) Discrepancy between gradiometer base size. 16.69 in Elekta, 16.80 in MNE
-#	2) Source of small z-coordinate offset (0.0003m). Not use in original SSS work,
-#	   but is present in Elekta's and MNE's coil definitions.
+#	2) Source of small z-coordinate offset (0.0003m) is due to manufacturing bug.
```
move from "to be sorted out" list

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -642,3 +685,116 @@ def _overlap_projector(data_int, data_res, corr):
+        temp_info = deepcopy(info)
+        for ch in temp_info['chs']:
+            ch['coil_type'] = pt_types[dir_ind]
+        coils_add, _, _, _ = _prep_meg_channels(temp_info, accurate=True,
```
can just do `_prep_meg_channels(...)[0]`

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -642,3 +685,116 @@ def _overlap_projector(data_int, data_res, corr):
+        S_in_add, S_out_add = _sss_basis(origin, coils_add, int_order,
+                                         ext_order)
+        # Scale spaces by gradiometer imbalance
+        S_in_all += S_in_add * imbalance[:, dir_ind][:, np.newaxis]
```
`... * imbalance[:, [dir_ind]]` is more compact if `imbalance` is small the array copy (instead of view) it forces might be worth the smaller code

Eric89GXL commented on mne/preprocessing/tests/test_maxwell.py:
```diff
@@ -28,12 +28,22 @@
     """Test multipolar moment and Maxwell filter"""
 
+    # Note: Elekta MaxFilter uses single precision, so expect filtered results
+    # to differ slightly.
```
just move this comment toward the top of the file, and remove it from both tests

Eric89GXL commented on mne/preprocessing/tests/test_maxwell.py:
```diff
@@ -253,4 +264,66 @@ def test_spatiotemporal_maxwell():
+
+    # Test 1D SSS fine calibration
+    raw_sss = maxwell_filter(raw, origin=[0., 0., 40.], int_order=int_order,
+                             ext_order=ext_order,
```
Only pass args that are not defaults. It helps ensure we don't change defaults in some bad way. So you should only need to pass `(raw, fine_cal_fname=fine_cal_fname)`.

Eric89GXL commented on mne/preprocessing/tests/test_maxwell.py:
```diff
@@ -253,4 +264,66 @@ def test_spatiotemporal_maxwell():
+    raw_sss = maxwell_filter(raw, origin=[0., 0., 40.], int_order=int_order,
+                             ext_order=ext_order,
+                             fine_cal_fname=fine_cal_fname)
+    assert_array_almost_equal(raw_sss._data[picks, :],
```
use public interface for data

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -47,6 +52,9 @@ def maxwell_filter(raw, origin=(0, 0, 40), int_order=8, ext_order=3,
     st_corr : float
         Correlation limit between inner and outer subspaces used to reject
         ovwrlapping intersecting inner/outer signals during spatiotemporal SSS.
+    fine_cal_fname : str | None
```
Let's just call this `fine_cal` and say it should be a filename in the description. Not worth the cost in characters to have `_fname` here IMO.

Eric89GXL commented on mne/forward/_make_forward.py:
```diff
@@ -288,17 +293,29 @@ def _prep_meg_channels(info, accurate=True, exclude=(), ignore_ref=False,
+        assert megcoils[0]['coord_frame'] == FIFF.FIFFV_COORD_HEAD
+        logger.info('MEG coil definitions created in head coordinates.')
+    else:
+        assert megcoils[0]['coord_frame'] == FIFF.FIFFV_COORD_DEVICE
```
I'm surprised this assertion passes since I thought in latest `master` the coord frame was always `head` (`transform=None` is assumed to be identity transform)

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -642,3 +685,116 @@ def _overlap_projector(data_int, data_res, corr):
+
+
+def _sss_basis_point(origin, info, int_order, ext_order, imbalance,
+                     head_frame=True):
```
Can you eliminate the `head_frame` changes altogether? I don't think you need them / use them at all, as `head_frame` is always `True` (right?).

wronk replied:
I added it for debugging purposes because it was a little easier for Samu to give me matrices for comparison in the device coordinate frame. I can change it back though

Eric89GXL commented:
I don't see anything obviously potentially wrong, other than the coordinate frame transform stuff, but I don't think those are even really wrong (see comments above).

Eric89GXL commented:
@wronk any changes to commit? I want to take a stab at the cross-talk. I can take care of my comments above while I'm at it.

wronk commented:
@Eric89GXL, you should be good to go. SNR is ~30 now, so still more work to do.

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -260,8 +296,9 @@ def _sss_basis(origin, coils, int_order, ext_order):
-        position info must be in the same coordinate frame as 'origin'
+        List of MEG coils. Each should contain coil information dict specifying
+        position, normals, weights, number of integration points and channel
+        type. All position info must be in the 'origin' coordinate frame
```
'origin' coord frame? You mean 'head' or 'device'? If it is 'origin' as in the origin of the SSS expansion, can you add a comment saying how these are related to `head` and `device`?

Eric89GXL commented on mne/preprocessing/maxwell.py:
```diff
@@ -642,3 +684,128 @@ def _overlap_projector(data_int, data_res, corr):
+
+            cal_chans.append(dict(ch=ch, ch_name=ch_name, loc=loc,
+                                  coil_trans=coil_trans,
+                                  calib_coeff=calib_coeff, ch_type=ch_type))
```
These are in device coords, correct? Please add a note. Are MEG channels in `info` typically in device coords, too? Mention if it's the same or different here.

wronk commented:
Update on all of this:
It seems like the sensor geometry (location and normal vectors) are being set correctly. Samu said we actually shouldn't overwrite sensor positions using the fine calibration file, but it has almost no effect on the final SNR, so we can come back to that issue later.

The geometry was really where I thought we were getting tripped up, so I'm getting more intermediate processing steps from Samu to continue pinning down the problem.

Eric89GXL commented:
Good to know the first part is working properly at least. Thanks for
sticking with it.

Eric89GXL commented:
@wronk can you add a warning when people use fine cal that it's experimental, since the shielding factors seem so similar? Then we can merge this, and make it work the MaxFilter way in a subsequent (non-monolithic) PR. We can also tackle the sensor-type warning in a different PR, since actually I think we should warn regardless of whether fine-cal is done (since we don't overwrite the channel info anymore). Sound good?

wronk commented:
Yep, sounds like a plan.

We do overwrite channel normals in fine-cal, just not channel location. Should I keep the warning for that reason?

Eric89GXL commented:
Yeah, we don't overwrite the channel location *or* the channel type (and thus the integration points that get used). So we should throw a warning even if fine cal is used.

Eric89GXL commented:
... so if you want to add the warning here, you can. But please refactor with the `fix_mag_coil_types` function to only search based on coil type in one function (DRY).

wronk commented:
Okay, I think I addressed all the above comments and added the documentation that we talked about a few days ago. Lmk if you want me to rename anything, but the functionality is there.

wronk commented:
AppVeyor timed out :(

Eric89GXL commented:
It's been doing that occasionally lately. Restarted.

agramfort commented:
travis is green

agramfort commented on mne/preprocessing/maxwell.py:
```diff
@@ -5,35 +6,54 @@
-def maxwell_filter(raw, origin=(0, 0, 40), int_order=8, ext_order=3,
-                   st_dur=None, st_corr=0.98, verbose=None):
+def maxwell_filter(raw, origin='default', int_order=8, ext_order=3,
+                   fine_cal=None, ctc=None, st_dur=None, st_corr=0.98,
```
how about 

fine_cal -> calibration
ctc -> crosstalk_correction
st_dur -> st_duration
st_corr -> st_correlation

?

wronk replied:
There are arguments for both ways I guess. Do we want the params to be clearer at first glance, but less like Elekta's naming and have long variable names to type out or vice versa?

agramfort replied:
python people like clear names :)

Eric89GXL commented:
Okay @agramfort ready for review/merge. We sorted out the issue with the difference in this commit:

https://github.com/mne-tools/mne-python/commit/156ffd6fffc61f5012f9eb099b4ea1a04d620b87

agramfort commented:
great job guys !

wronk commented:
Niiiice!
